viewing_party/party.py

```diff
@@ -1,23 +1,123 @@
+def get_watched_avg_rating(user_data):
+    sum = 0 
+
+    if len(user_data["watched"]) == 0:
+        return 0.0
+    
+    for movie in user_data["watched"]:
+        sum += movie["rating"] 
+
+    return sum / len(user_data["watched"]) 
 
+def get_most_watched_genre(user_data):
+    if len(user_data["watched"]) == 0:
+        return None
+    
+    genre_dictionary = {}
+    
+    for movie in user_data["watched"]:
+        if movie["genre"] not in genre_dictionary:
+            genre_dictionary[movie["genre"]] = 1
+        else:
+            genre_dictionary[movie["genre"]] += 1
+
+    # gets max value from dictioanry and returns the key
+    max_genre = max(genre_dictionary, key = genre_dictionary.get)
+
+    return max_genre
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def get_friend_watched_list(user_data):
+    friend_watched_list = []
+    for friend in user_data["friends"]:
+        for friend_movie in friend["watched"]:
+            friend_watched_list.append(friend_movie)
+    return friend_watched_list
+
+
+def get_unique_watched(user_data):
+    friend_list = get_friend_watched_list(user_data)
+    unique_list = []
+
+    for movie in user_data["watched"]:
+        if movie not in friend_list:
+            unique_list.append(movie)
+
+    return unique_list
+
+
+def get_friends_unique_watched(user_data):
+    friend_list = get_friend_watched_list(user_data)
+    unique_list = []
+
+    for movie in friend_list:
+        if movie not in user_data["watched"] and movie not in unique_list:
+            unique_list.append(movie)
+    return unique_list
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
+    friend_watched_list = (get_friends_unique_watched(user_data))
+    recommended_movie_list = []
+    
+    for movie in friend_watched_list:
+        if movie["host"] in user_data["subscriptions"]:
+            recommended_movie_list.append(movie)
+
+    return recommended_movie_list
 
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+def get_new_rec_by_genre(user_data):
+    most_watched_genre = get_most_watched_genre(user_data)
+    friend_watched_list = (get_friends_unique_watched(user_data))
+    recommended_movie_list = []
+
+    for movie in friend_watched_list:
+        if movie["genre"] == most_watched_genre:
+            recommended_movie_list.append(movie)
+    
+    return recommended_movie_list
+
+def get_rec_from_favorites(user_data):
+    user_watched_list = get_unique_watched(user_data)
+    recommended_movie_list = []
+
+    for movie in user_data["favorites"]:
+        if movie in user_watched_list:
+            recommended_movie_list.append(movie)
+    
+    return recommended_movie_list

```

Comment:
Great Job! You both did an amazing job with this project! Please make sure to reach out if you have any questions about the comments I left. Also, if you are wondering about why I didn't leave many comments about `Time Complexity` is that due to the nature of this project and the placement of it I felt as if it would be be unfair since this is you all's first time working with nested loops and data structures.  ⭐️ Though you all kept it generally in the `0(n^2)` range which is acceptable for this project given there's no way to really get around that given the data structures that you are working on.